csrc/cpu/comm/ccl.cpp

```diff
@@ -193,6 +450,70 @@ void all_reduce_caching(torch::Tensor& data,
                  .wait());
 }
 
+void inference_allreduce(torch::Tensor& data, py::object op, py::object group, bool async_op)
+{
+    static py::object ReduceOp = py::module_::import("deepspeed.comm").attr("ReduceOp");
+    static auto ReduceOpSum = (int)py::int_(ReduceOp.attr("SUM").attr("value"));
+
+    assert(py::int_(op.attr("value")) == ReduceOpSum);
+
+    auto numel = data.numel();
+
+    int data_size = 0;
+    bool data_type_fallback = false;
+
+    switch (data.scalar_type()) {
+        case c10::ScalarType::BFloat16: data_size = numel * 2; break;
+        case c10::ScalarType::Float: data_size = numel * 4; break;
+        default: data_type_fallback = true;
+    }
+
+    if (data_size > MAX_BUF_SIZE || (numel % 16) != 0 || data_type_fallback || !all_ranks_local_p) {
```

Comment:
Would be good to replace `16` with a descriptive symbolic constant. 
I assume this is some alignment constraint. If so, I am surprised the same value is used for both bf16 and fp32.

Reply:
Thanks!  Added a new constant `VECTOR_LENGTH_IN_BYTES` and define it to 32.  This would reflect the length of vector used in reduce kernels (256bits=>32bytes).   Will check with `data_size % VECTOR_LENGTH_IN_BYTES` here, and also use the same constant to control loop steps in reduce kernels.